### PR TITLE
Pydantic latest

### PR DESCRIPTION
**Pull request by joamatab**
fixes https://github.com/gdsfactory/gdsfactory/issues/1591

### CODE REVIEW

**Comment by dmontagu** [comment]
Hey there, happy to help with this. I tried setting up a dev environment and had some issues, do I need to use conda or something? I'm currently getting
```python
gdsfactory on  main [?] is 📦 v6.90.7 via 🐍 v3.10.5 (env3.10) 
❯ gf install klayout-integration

Traceback (most recent call last):
  File "/Users/davidmontague/Programming/pydantic/gdsfactory/env3.10/bin/gf", line 5, in <module>
    from gdsfactory.cli import cli
  File "/Users/davidmontague/Programming/pydantic/gdsfactory/gdsfactory/__init__.py", line 23, in <module>
    from gdsfactory.component_layout import Group
  File "/Users/davidmontague/Programming/pydantic/gdsfactory/gdsfactory/component_layout.py", line 12, in <module>
    from gdstk import Polygon
ImportError: dlopen(/Users/davidmontague/Programming/pydantic/gdsfactory/env3.10/lib/python3.10/site-packages/gdstk.cpython-310-darwin.so, 0x0002): symbol not found in flat namespace '__ZN5gdstk12error_loggerE'
```

I tried using the dev container, after running the setups in the README.md on it after creation, I was at least able to run pytest, but I got:

```
/offset_d3fc76c5.gds')
FAILED tests/test_paths.py::test_gds[test_path] - AssertionError: Reference GDS file for 'test_path' not found. Writing to PosixPath('/workspaces/gdsfactory/gdslib/gds/gds_ref/test_path.gds')
FAILED tests/test_paths.py::test_gds[rename] - AssertionError: Reference GDS file for 'rename' not found. Writing to PosixPath('/workspaces/gdsfactory/gdslib/gds/gds_ref/rename.gds')
FAILED tests/test_paths.py::test_gds[transition] - AssertionError: Reference GDS file for 'transition' not found. Writing to PosixPath('/workspaces/gdsfactory/gdslib/gds/gds_ref/transition.gds')
FAILED tests/test_route_ports_to_side.py::test_sample_route_sides - AssertionError: Reference GDS file for '_sample_route_sides' not found. Writing to PosixPath('/workspaces/gdsfactory/gdslib/gds/gds_ref/_sample_route_sides.gds')
======================================================== 326 failed, 852 passed, 174 warnings in 105.64s (0:01:45) ========================================================
@dmontagu ➜ /workspaces/gdsfactory (main) $ 
```
Is there something else I need to do to set up a dev environment? (Sorry for the spam on your PR)

File: gdsfactory/cross_section.py [hunk 1]
```diff
@@ -175,7 +175,7 @@ class Config:
         """Configuration."""
 
         extra = "forbid"
-        fields = {
+        Annotated = {
```
Comment (dmontagu) [inline gdsfactory/cross_section.py:178]: Was there something in our docs that made you want to change `fields` to `Annotated`? Fortunately I think these can all be handled by moving the `exclude=True` onto the `= Field` bits above (I'll open a PR against this branch doing that), but I'm wondering if there was something you saw that inspired this change.

**Comment by dmontagu** [comment]
I see this in your test_docs CI job:
```
INFO: pip is looking at multiple versions of sax to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install gdsfactory and gdsfactory[database,devsim,meow,ray,sax,tidy3d]==6.90.2 because these package versions have conflicting dependencies.

The conflict is caused by:
    gdsfactory[database,devsim,meow,ray,sax,tidy3d] 6.90.2 depends on pydantic==2.0a3
    sax 0.8.6 depends on pydantic<2
ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict
```
We are planning to release a package called `pydantic_next` on PyPI that you can use instead of `pydantic>=2.0` so that it will let other dependencies of yours wait to upgrade, will let you know once that's released.

**Comment by daquintero** [comment]
Hey guys, cheers for all the work you are doing! I love GDSFactory and Pydantic.

I thought I couldn't quite wait for this to be implemented as I needed a v2-compatible environement sooner, so I gave it a go at porting it - although I've given up now as it's a massive job. Maybe it can be used as a reference for changes that might be required as there were quite a lot. Also v2 is incompatible with 1.1 so backwards compatibility might be an issue. I am aware of the dependencies issue and `pydantic_next` might be the way for integration but even then there's a big refactor coming.

In case it is useful though here is my forked patched branch https://github.com/daquintero/gdsfactory/tree/pydantic_v2_test_broken

**Comment by joamatab** [comment]
Thank you Dario,

I created a PR based on your work
https://github.com/gdsfactory/gdsfactory/pull/1658

so we can continue from there

**Pull request closed without merging**